### ISSUE TITLE
Different webpack.config.js pickup logic, than original webpack-cli

### ISSUE THREAD

**Issue by just-boris** [webpack-contrib/webpack-serve#65]
This issue is for a*:

- [x] **bug**
- [x] **feature** request
- [ ] **modification** request

**not sure, whether the current behavior is qualified like a bug or a missing feature*

### Description

Webpack-cli package picks up `webpack.config.js` file [using `interpret` module](https://github.com/webpack/webpack-cli/blob/master/bin/convert-argv.js#L4). It allows to pick up files with different file extensions and process them when appropriate. For example, I have `webpack.config.ts` file and it works with latest Webpack 4 perfectly.

As I see, webpack-serve [uses different `cosmiconfig` module](https://github.com/webpack-contrib/webpack-serve/blob/master/cli.js#L72), which lacks that functionality, `webpack-serve` simply ignores the file and attempts to build files in zero-config mode.

### Expected Behavior

If config is understood by `webpack-cli` it should work with `webpack-serve` too.

### Actual Behavior

`webpack` command runs well, but `webpack-serve` fails, because it doesn't see the config

### New Feature Use Case

Having both packages working the same way would improve dev-experience and will help to avoid confusion.

---

P.S. there was a similar issue https://github.com/webpack-contrib/webpack-serve/issues/51 already, but here I am reporting more general issue, rather than a particular use-case.

**Comment by rubencodes** [comment]
I think this is related: I had my `webpack.config.babel.js` with `import`'s and I noticed that didn't work. Had to revert to `require`. Would love for that to be supported as well.

**Comment by shellscape** [comment]
@just-boris thanks for the issue 🍺  Unfortunately the maintainers of this module disagree with how `webpack-cli` functions, and believe it's packed with far too much bloat. `webpack-serve` isn't `webpack-cli`. If you're looking for something more aligned with what the CLI let's you get away with, then `webpack-dev-server` might be more your flavor.

This module will not implement TypeScript, Babel, nor any other preprocessor support. You must feed it regular JavaScript.

**Comment by just-boris** [comment]
As [webpack-dev-server states in their readme](https://github.com/webpack/webpack-dev-server#project-in-maintenance):

> Please note that webpack-dev-server is presently in a maintenance-only mode [...] Use webpack-serve for a fast alternative.

I thought that `webpack-serve` is more modern dev-server implementation, that should be preferred for future projects. Turns out, that it is not, and two dev-servers will exist and I will need to choose between them.

Thank you for the quick update @shellscape!

**Comment by shellscape** [comment]
> Turns out, that it is not, two dev-servers will exist and I will need to choose between them.

Well that's just patently false. And I want to make sure that people reading this thread understand that. Please don't make silly false statements like that. Perhaps you're just too acclimated to projects that aren't managed well and will implement anything and everything that users ask for.

Or as an alternative, propose a solution that _doesn't include_ adding bloat nor specific preprocessor support to the project.

**Comment by shellscape** [comment]
And to follow up on this further:

One of the big differences between this project and `webpack-dev-server` is the fact that the CLI for this project is aligned with the API. That means you can use the API as easy as the CLI, and the CLI is only a thin wrapper around the API. That also means that if you need specific preprocessor support, it's trivial to write a script to run `webpack-serve` from npm scripts in whatever language flavor you want to. And that doesn't require special support from the module to do.

If passers-by would like reading material on why what `webpack-cli` does is _bad_ you can read about why packages like [mocha](https://github.com/mochajs/mocha/wiki/compilers-deprecation) don't directly support specific compilers, and instead allow users on the CLI to specify packages to `require` during execution (which invariably just hook `require` to allow processor/transpiler requiring). But that can also be done if the main entry file just requires that as well.

 Basically this amounts to: _don't expect a module to cater to every special case, and don't be lazy when you want special sauce._

**Comment by just-boris** [comment]
> Please don't make silly false statements like that. 

I am sorry, I just wanted to be clear in my understanding of relationships between different pieces of Webpack ecosystem. Do I understand correctly now, that `webpack-serve` *is* the recommended tool for new projects?

> Or as an alternative, propose a solution that doesn't include adding bloat nor specific preprocessor support to the project.

[`interpret`](https://github.com/js-cli/js-interpret) (a package, that webpack-cli uses) is a dependency-free package. It just contains a map "extension"-"module loader", that other tools can use. The eventual responsibility to install a preprocessor is on end user. It takes [a few lines](https://github.com/webpack/webpack-cli/blob/master/bin/convert-argv.js#L102-L117) in webpack-cli to handle this. Of, course duplicating those lines in `webpack-serve` would not help a lot, but maybe it is possible to create a shared package (`load-webpack-config`?) that always returns a proper configuration according to specified rules.

> it's trivial to write a script to run webpack-serve from npm scripts in whatever language flavor you want to

This is one of the possible options, that I will probably apply right now. Thank you for the hint!

> don't expect a module to cater to every special case, and don't be lazy when you want special sauce.

I am fine with having more code on my side if needed, but it is not good when two very related parts of typical Webpack setup behave differently.

**Comment by polkovnikov-ph** [comment]
> trivial to write a script to run webpack-serve from npm scripts in whatever language flavor you want to

If it's that trivial, why there is no mention in documentation on how to do that? From the look of [this file](https://github.com/webpack-contrib/webpack-serve/blob/master/index.js) it's in no way clear where TypeScript support should go.

**Comment by shellscape** [comment]
@polkovnikov-ph because it's basic info that a developer at any level with minimal critical thinking ability should be able to glean. the API docs are there clear as day and the CLI serves as a functional example of how to use the API. Not to mention all the tests that are there for reference as well. Writing a simple top-level script file for Node is _a basic skill_. If one has waded into Node and doesn't have that skill, it's time to take a step back and get back to basics. And that applies to TypeScript, Flow, CoffeeScript, ES6+Babel, etc.

**Comment by polkovnikov-ph** [comment]
@shellscape After reading all the rant, I still didn't see a link to `@types/webpack-serve`, which is kinda required to do that trivial `serve({config})`. Oh, wait, it doesn't exist.

**Comment by shellscape** [comment]
@polkovnikov-ph I would encourage you to use a different module that meets your expectations and caters to your particular needs. This is not it.

**Comment by shellscape** [comment]
The `--require` feature has now been added in #83.
